Add Table support to Engine

trgibeau opened the pull request:
Enabled the ability to access tables from the engine

- Added Try/EnableTable(TableDescriptor tableDescriptor)
-- Allows callers to enable the table before calling Process()
- Added IsTableDataAvailable(TableDescriptor tableDescriptor)
-- Allows callers to check if the table has data after calling Process()
- Added Try/BuildTable(TableDescriptor tableDescriptor)
-- Allows callers to build a table after calling Process
- Added Unit and Functional tests

jmaxson-ms commented on src/Microsoft.Performance.Toolkit.Engine/ITableResult.cs:
```diff
@@ -0,0 +1,24 @@
+
+namespace Microsoft.Performance.Toolkit.Engine
+{
+    public interface ITableResult
```
we'll need to have doc comments on this since it's public :)

jmaxson-ms replied:
Sorry, one more doc comment on the interface itself

anramanath commented on src/Microsoft.Performance.Toolkit.Engine.Tests/TestData/EngineBuildTabeTestDtos.cs:
```diff
@@ -0,0 +1,51 @@
+namespace Microsoft.Performance.Toolkit.Engine.Tests.TestData
+{
+    [DataContract]
+    public class EngineBuildTableTestSuiteDto
```
What is Dto?

jmaxson-ms replied:
https://en.wikipedia.org/wiki/Data_transfer_object

jmaxson-ms commented:
Going to have to merge Zack's IDispose change on the Engine.

mslukebo commented on src/Microsoft.Performance.Toolkit.Engine/Engine.cs:
```diff
@@ -433,6 +451,56 @@ public bool TryEnableCooker(DataCookerPath dataCookerPath)
+        /// </param>
+        public void EnableTable(TableDescriptor descriptor)
+        {
+            this.ThrowIfProcessed();
```
can we add a <throws> to the docstring

mslukebo commented on src/Microsoft.Performance.Toolkit.Engine/Engine.cs:
```diff
@@ -433,6 +451,56 @@ public bool TryEnableCooker(DataCookerPath dataCookerPath)
+
+            if (!this.TryEnableTable(descriptor))
+            {
+                throw new TableNotFoundException(descriptor);
```
and one for this

mslukebo commented on src/Microsoft.Performance.Toolkit.Engine/EngineExecutionResults.cs:
```diff
@@ -41,10 +46,12 @@ public sealed class RuntimeExecutionResults
         ///     - or -
         ///     <paramref name="repository"/> is <c>null</c>.
         /// </exception>
+        /// TODO: Going to move RuntimeExecutionResults to internal and expose calls via new interface
```
can we denote this with a `<remarks>` (or even an `ObsoleteAttribute`)

trgibeau replied:
I'm going to do this in a different PR (didn't want it to get too big :))

It will be before we snap 0.109

mslukebo commented on src/Microsoft.Performance.Toolkit.Engine/EngineExecutionResults.cs:
```diff
@@ -215,5 +223,217 @@ public bool TryQueryOutput<T>(DataOutputPath dataOutputPath, out T data)
+        ///     <c>true</c> the table has data available;
+        ///     <c>false</c> otherwise.
+        ///     <c>null</c> the table does not support checking if data is available, 
+        ///     recommend to call <see cref="BuildTable(TableDescriptor)"/>.
```
what does this mean

trgibeau replied:
IsDataAvailable is not required to be implemented. 

Should reword a little?

mslukebo replied:
the part I was confused on is "recommend to call <see..."

trgibeau replied:
Ah, wanted to say that the caller should call BuildTable directly without needing to check if it has data

mslukebo commented on src/Microsoft.Performance.Toolkit.Engine/EngineExecutionResults.cs:
```diff
@@ -215,5 +223,217 @@ public bool TryQueryOutput<T>(DataOutputPath dataOutputPath, out T data)
+        ///     The built table.
+        /// </returns>
+        /// <exception cref="TableNotBuiltException">
+        ///     Table cannot be built for the given <paramref name="tableDescriptor"/>.
```
nit: A table

mslukebo commented on src/Microsoft.Performance.Toolkit.Engine/EngineExecutionResults.cs:
```diff
@@ -215,5 +223,217 @@ public bool TryQueryOutput<T>(DataOutputPath dataOutputPath, out T data)
+        /// </exception>
+        public ITableResult BuildTable(TableDescriptor tableDescriptor)
+        {
+            if(TryBuildTable(tableDescriptor, out ITableResult filledTableBuilder))
```
nit: space after if

trgibeau commented:
> Going to have to merge Zack's IDispose change on the Engine.

Thank Jayson, just remerged and added dispose support

@zachnew-msft can you double check that I added dispose in the right places

jmaxson-ms commented on src/Microsoft.Performance.Toolkit.Engine/Engine.cs:
```diff
@@ -48,6 +50,9 @@ public abstract class Engine
         private ReadOnlyCollection<DataCookerPath> enabledCookersRO;
 
+        private readonly List<TableDescriptor> enabledTables;
+        private readonly ReadOnlyCollection<TableDescriptor> enabledTablesRO;
```
these shouldn't be readonly because they need to be nulled in Dispose

jmaxson-ms commented on src/Microsoft.Performance.Toolkit.Engine/Engine.cs:
```diff
@@ -269,6 +278,30 @@ private set
+        /// <summary>
+        ///     Gets the collection of all enabled tables
+        /// </summary>
+        public IEnumerable<TableDescriptor> EnableTables
```
>EnableTables [](start = 44, length = 12)

typo: EnabledTables

jmaxson-ms commented on src/Microsoft.Performance.Toolkit.Engine/Engine.cs:
```diff
@@ -269,6 +278,30 @@ private set
 
+        /// <summary>
+        ///     Gets the collection of all enabled tables
+        /// </summary>
```
for each of the new methods/properties, this should be added to the doc comments:

        /// <exception cref="ObjectDisposedException">
        ///     This instance is disposed.
        /// </exception>

jmaxson-ms commented on src/Microsoft.Performance.Toolkit.Engine/Engine.cs:
```diff
@@ -598,6 +631,64 @@ public bool TryEnableCooker(DataCookerPath dataCookerPath)
+        ///     <c>false</c> otherwise. Note that <c>false</c> is
+        ///     always returned when <see cref="IsProcessed"/> is <c>true</c>.
+        /// </returns>
+        public bool TryEnableTable(TableDescriptor descriptor)
```
>descriptor [](start = 51, length = 10)

should we Guard.NotNull(descriptor)? if so, we should also add the ArgumentNullException to the doc comments (and bubbled up to EnableTable)

jmaxson-ms commented on src/Microsoft.Performance.Toolkit.Engine/Engine.cs:
```diff
@@ -863,10 +956,38 @@ private RuntimeExecutionResults ProcessCore()
+                {
+                    executor.Processor.EnableTable(table);
+                    processorTables.Add(table, executor.Processor);
+                }
```
I'm wondering if this should be part of an "else" block below the if block below. Given the a table is uniquely identified by a Guid, it would be a bug to be in both sets.

jmaxson-ms commented:
>         ///     Gets any non-fatal errors that occured during engine

would you mind fixing this typo while you're modifying this file?

---
Refers to: src/Microsoft.Performance.Toolkit.Engine/Engine.cs:259 in 02d70c2. [](commit_id = 02d70c2b6f25ba990013c9e0bc8bcfec59413e4d, deletion_comment = False)

jmaxson-ms commented on src/Microsoft.Performance.Toolkit.Engine/EngineExecutionResults.cs:
```diff
@@ -41,10 +46,12 @@ public sealed class RuntimeExecutionResults
             IDataExtensionRetrievalFactory retrievalFactory,
-            IDataExtensionRepository repository)
+            IDataExtensionRepository repository,
+            IDictionary<TableDescriptor, ICustomDataProcessor> tableToProcessorMap)
```
>tableToProcessorMap [](start = 63, length = 19)

Guard.NotNull and update doc comment for ArgumentNullException

trgibeau replied:
This is going to moved to internal in another PR

trgibeau replied:
Added Guard.NotNull

jmaxson-ms commented on src/Microsoft.Performance.Toolkit.Engine/EngineExecutionResults.cs:
```diff
@@ -215,5 +223,222 @@ public bool TryQueryOutput<T>(DataOutputPath dataOutputPath, out T data)
+                }
+                catch
+                {
+                    return false;
```
should we report some error to the user in this case? (same in TryOnProcessor) if we want to , we might have to let the exception bubble up to where we know what the func is actually doing

jmaxson-ms replied:
gah, just realized this is in a Try[Method]. not sure what the right thing to do here is

---
In reply to: [592041314](https://github.com/microsoft/microsoft-performance-toolkit-sdk/pull/56#discussion_r592041314) [](ancestors = 592041314)

jmaxson-ms replied:
maybe just have a TryGetOn[Repository|Processor], and not pass in the func?

---
In reply to: [592041488](https://github.com/microsoft/microsoft-performance-toolkit-sdk/pull/56#discussion_r592041488) [](ancestors = 592041488,592041314)

jmaxson-ms commented on src/Microsoft.Performance.Toolkit.Engine/EngineExecutionResults.cs:
```diff
@@ -20,6 +24,7 @@ public sealed class RuntimeExecutionResults
         private readonly IDataExtensionRetrievalFactory retrievalFactory;
         private readonly IDataExtensionRepository repository;
         private readonly IEnumerable<DataCookerPath> sourceCookers;
+        private readonly IDictionary<TableDescriptor, ICustomDataProcessor> tableToProcessorMap;
```
>ICustomDataProcessor [](start = 54, length = 20)

is it possible for this Table to exist in more than one instance of the same ICustomDataProcessor?

jmaxson-ms replied:
An engine is a single session, so this shouldn't be an issue.

---
In reply to: [592042897](https://github.com/microsoft/microsoft-performance-toolkit-sdk/pull/56#discussion_r592042897) [](ancestors = 592042897)

trgibeau replied:
nope :)

jmaxson-ms commented on src/Microsoft.Performance.Toolkit.Engine/Engine.cs:
```diff
@@ -863,10 +967,40 @@ private RuntimeExecutionResults ProcessCore()
+                    foreach (var executor in executors.Where(x => x.Context.CustomDataSource.AvailableTables.Contains(table)))
+                    {
+                        executor.Processor.EnableTable(table);
+                        processorTables.Add(table, executor.Processor);
```
based on what I remember @zachnew-msft said today... an executor is tied to a CDS, and a table should not exist in multiple custom data sources, so I think semantically it would make more sense to do a .First() instead of a foreach.

clearly not a functional change, just seems more clear to me what we actually expect to happen if we did it that way

trgibeau replied:
Tweaked it to be "Single" since it should be there

jmaxson-ms commented on src/Microsoft.Performance.Toolkit.Engine/EngineException.cs:
```diff
@@ -189,6 +189,172 @@ public override string ToString()
+        ///     The error that is the cause of this error.
+        /// </param>
+        public TableException(TableDescriptor tableDescriptor, Exception inner)
+            : base($"The requested table '{tableDescriptor}' was not found", inner)
```
>$"The requested table '{tableDescriptor}' was not found" [](start = 19, length = 56)

this should be updated

jmaxson-ms commented on src/Microsoft.Performance.Toolkit.Engine/EngineExecutionResults.cs:
```diff
@@ -215,5 +225,228 @@ public bool TryQueryOutput<T>(DataOutputPath dataOutputPath, out T data)
+        }
+
+        // Will perform the Func on the Repository if the table is contained, else false.
+        private bool FuncOnRepositoryIfContained<TResult>(TableDescriptor tableDescriptor, Func<ITableExtensionReference, IDataExtensionRetrieval, TResult> func, out TResult result)
```
>FuncOnRepositoryIfContained [](start = 21, length = 27)

My personal preference is to avoid types in names - so something like "ExecuteOnRepositoryIfContained"... but not a big deal with way. Feel free to ignore :)

jmaxson-ms commented on src/Microsoft.Performance.Toolkit.Engine/EngineException.cs:
```diff
@@ -189,6 +189,172 @@ public override string ToString()
+    ///     <see cref="ITableResult"/> that with a specified <see cref="TableDescriptor"/>.
+    /// </summary>
+    [Serializable]
+    public class TableException
```
>TableException [](start = 17, length = 14)

I wonder if we should require a message on this exceptions as it is fairly generic

jmaxson-ms reviewed (approved):
:shipit:

jmaxson-ms commented on src/Microsoft.Performance.Toolkit.Engine/EngineExecutionResults.cs:
```diff
@@ -215,5 +225,229 @@ public bool TryQueryOutput<T>(DataOutputPath dataOutputPath, out T data)
+            }
+            catch (Exception inner)
+            {
+                throw new TableException($"An excpetion was thrown while calling IsDataAvailable for the {tableDescriptor}.", tableDescriptor, inner);
```
>excpetion [](start = 46, length = 9)

typo

jmaxson-ms commented on src/Microsoft.Performance.Toolkit.Engine/EngineExecutionResults.cs:
```diff
@@ -215,5 +225,229 @@ public bool TryQueryOutput<T>(DataOutputPath dataOutputPath, out T data)
+                innerException = inner;
+            }
+
+            throw new TableException($"An excpetion was thrown while calling BuildTable for the {tableDescriptor}.", tableDescriptor, innerException);
```
>excpetion [](start = 42, length = 9)

copy/paste

jmaxson-ms commented on src/Microsoft.Performance.Toolkit.Engine/EngineException.cs:
```diff
@@ -189,6 +189,178 @@ public override string ToString()
+        : EngineException
+    {
+        /// <inheritdoc />
+        public TableException() : base() { }
```
>public [](start = 8, length = 6)

want this private?

trgibeau replied:
I think since its serializable it needs to be public
